File: src/main/java/meowing/zen/utils/EventBus.java
package meowing.zen.utils;

import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

public class EventBus {
    private static final Map<Class<?>, Map<Object, Consumer<?>>> listeners = new HashMap<>();
    
    public static <T> void register(Class<T> eventType, Object owner, Consumer<T> handler) {
        listeners.computeIfAbsent(eventType, k -> new HashMap<>()).put(owner, handler);
    }
    
    public static void unregister(Object owner) {
        listeners.values().forEach(map -> map.remove(owner));
    }
    
    @SuppressWarnings("unchecked")
    public static <T> void fire(T event) {
        Map<Object, Consumer<?>> handlers = listeners.get(event.getClass());
        if (handlers != null) {
            handlers.values().forEach(handler -> ((Consumer<T>) handler).accept(event));
        }
    }
}

File: src/main/kotlin/meowing/zen/utils/EventBus.kt
package meowing.zen.utils

import java.util.concurrent.ConcurrentHashMap
import java.util.function.Consumer

object EventBus {
    private val listeners = ConcurrentHashMap<Class<*>, ConcurrentHashMap<Any, Consumer<*>>>()

    @JvmStatic
    fun <T> register(eventType: Class<T>, owner: Any, handler: Consumer<T>) {
        listeners.computeIfAbsent(eventType) { ConcurrentHashMap() }[owner] = handler
    }

    @JvmStatic
    fun unregister(owner: Any) {
        listeners.values.forEach { map -> map.remove(owner) }
    }

    @JvmStatic
    @Suppress("UNCHECKED_CAST")
    fun <T> fire(event: T) {
        val handlers = listeners[event!!::class.java]
        if (handlers != null && handlers.isNotEmpty()) {
            handlers.values.parallelStream().forEach { handler ->
                try {
                    (handler as Consumer<T>).accept(event)
                } catch (ignored: Exception) {
                }
            }
        }
    }
}